app/ghcup/BrickMain.hs

```diff
@@ -59,70 +74,243 @@ import qualified Data.Text.Lazy.Builder        as B
+
+-- | Handle events for list cursor movement.  Events handled are:
+--
+-- * Up (up arrow key). If first element of section, then jump prev section
+-- * Down (down arrow key). If last element of section, then jump next section
+-- * Page Up (PgUp)
+-- * Page Down (PgDown)
+-- * Go to next section (Tab)
+-- * Go to prev section (BackTab)
+handleGenericListEvent :: (Foldable t, L.Splittable t, Ord n)
+                       => BrickEvent n a
+                       -> EventM n (GenericSectionList n t e) ()
+handleGenericListEvent (VtyEvent (Vty.EvResize _ _))              = pure ()
+handleGenericListEvent (VtyEvent (Vty.EvKey (Vty.KChar '\t') [])) = sectionListFocusRingL %= F.focusNext
+handleGenericListEvent (VtyEvent (Vty.EvKey Vty.KBackTab []))     = sectionListFocusRingL %= F.focusPrev
+handleGenericListEvent (MouseDown _ Vty.BScrollDown _ _)          = moveDown
+handleGenericListEvent (MouseDown _ Vty.BScrollUp _ _)            = moveUp
+handleGenericListEvent (VtyEvent (Vty.EvKey Vty.KDown []))     = moveDown
+handleGenericListEvent (VtyEvent (Vty.EvKey Vty.KUp []))       = moveUp
+handleGenericListEvent (VtyEvent ev) = do
+    ring <- use sectionListFocusRingL
+    case F.focusGetCurrent ring of
+        Nothing -> pure ()
+        Just l  -> Brick.zoom (toLensVL $ sectionL l) $ L.handleListEvent ev
+handleGenericListEvent _ = pure ()
+
+-- This re-uses Brick.Widget.List.renderList
+renderSectionList :: (Traversable t, Ord n, Show n, Eq n, L.Splittable t)
+                  => (Bool -> n -> Widget n)             -- ^ Rendering function for separator between sections. True for selected before section
+                  -> (Bool -> n -> Widget n -> Widget n) -- ^ Rendering function for the borders. True for selected section
+                  -> (Bool -> e -> Widget n)             -- ^ Rendering function of the list element, True for the selected element
+                  -> Bool                                -- ^ Whether the section list has focus
+                  -> GenericSectionList n t e            -- ^ The section list to render
+                  -> Widget n
+renderSectionList render_separator render_border render_elem section_focus (GenericSectionList focus elms sl_name) =
+  Widget Greedy Greedy $ render $
+    Brick.viewport sl_name Vertical $
+      V.foldl' (\wacc list ->
+                  let has_focus_list = is_focused_section list
+                      list_name = L.listName list
+                  in   wacc
+                    <=> render_separator has_focus_list list_name
+                    <=> inner_widget has_focus_list list_name list
+              )
+              Brick.emptyWidget elms
+  where
+    is_focused_section l = section_focus && Just (L.listName l) == F.focusGetCurrent focus
+    inner_widget has_focus k l = 
+      let w = render_border has_focus k (Brick.vLimit (length l) $ L.renderList render_elem has_focus l)
+       in if has_focus then Brick.visible w else w
 
```

Comment:
rendering uses [`Brick.Widget.List.renderList`](https://hackage.haskell.org/package/brick-1.10/docs/Brick-Widgets-List.html#v:renderList) instead of define it ourself. The separators between sections are provided in the first argument. Previously, this was done by hand. Notice that the `viewport` follows the section, not the element in the list. This is a behavioural change wrt current `master`

Btw, the second argument is currently empty, but allow to define borders (not separators) to each section.